src/new-client/src/App/LiveRates/Tile/state.ts

```diff
@@ -0,0 +1,138 @@
+import { getCurrencyPair$ } from "services/currencyPairs"
+import { emitTooLongMessage } from "utils/emitTooLong"
+import { bind } from "@react-rxjs/core"
+
+// Notional
+const DEFAULT_NOTIONAL = 1_000_000
+
+const [rawNotional$, onChangeNotionalValue] = createListener<{
+  symbol: string
+  value: string
+}>()
+
+const mapNotionals$ = rawNotional$.pipe(
+  split(
+    (e) => e.symbol,
+    (rawNotional$) =>
+      rawNotional$.pipe(
+        pluck("value"),
+        filter((value) => !Number.isNaN(Number(value))),
+      ),
+  ),
+  collect(),
+)
+mapNotionals$.subscribe()
+
+const [useNotional, getNotional$] = bind(
+  (symbol: string) =>
+    mapNotionals$.pipe(
+      exhaustMap((map) => (map.has(symbol) ? map.get(symbol)! : EMPTY)),
+    ),
+  DEFAULT_NOTIONAL.toString(10),
+)
+
+export { onChangeNotionalValue, useNotional }
+
+// Execution
+const [tileExecutions$, sendExecution] = createListener<{
+  symbol: string
+  direction: Direction
+}>()
+export { sendExecution }
+
+const mapExecutions$ = tileExecutions$.pipe(
+  split((e) => e.symbol),
+  collect(),
+)
+
+// Dismiss Message
+const DISMISS_TIMEOUT = 5_000
+const [dismiss$, onDismissMessage] = createListener<string>()
+export { onDismissMessage }
+
+// TileState
+export enum TileStates {
+  Ready,
+  Started,
+  TooLong,
+  Timeout,
+  Finished,
+}
+
+interface NoTradeState {
+  status:
+    | TileStates.Ready
+    | TileStates.Started
+    | TileStates.TooLong
+    | TileStates.Timeout
+  trade?: undefined
+}
+
+interface TradeState {
+  status: TileStates.Finished
+  trade: ExecutionTrade
+}
+
+export type TileState = TradeState | NoTradeState
+
+let nextId = 1
+const getId = () => (nextId++).toString()
+const TAKING_TOO_LONG = 2_000
+
+export const [useTileState] = bind(
+  (symbol: string) =>
+    mapExecutions$.pipe(
+      exhaustMap((map) => (map.has(symbol) ? map.get(symbol)! : EMPTY)),
+      withLatestFrom(
+        getNotional$(symbol),
+        getPrice$(symbol),
+        getCurrencyPair$(symbol),
+      ),
+      map(([{ direction }, notional, price, { base, terms, symbol }]) => ({
+        id: getId(),
+        currencyPair: symbol,
+        dealtCurrency: direction === Direction.Buy ? base : terms,
+        direction,
+        notional: Number(notional),
+        spotRate: direction === Direction.Buy ? price.ask : price.bid,
+      })),
+      exhaustMap((execution) =>
+        concat(
+          execute$(execution).pipe(
+            map((trade) =>
+              trade.status === ExecutionStatus.Timeout
+                ? { status: TileStates.Timeout }
+                : { status: TileStates.Finished, trade },
+            ),
+            emitTooLongMessage(TAKING_TOO_LONG, {
+              status: TileStates.TooLong,
+            }),
+            startWith({ status: TileStates.Started }),
+          ),
+          race([
+            dismiss$.pipe(
+              filter((s) => s === symbol),
+              take(1),
```

Comment:
Still kind of confused why take(1) fixed the bug, it seems even the dismiss$ doesn't complete, it will still return a status: ready, right? Why it will make our buttons deactivated?

Reply:
The reason is that:

- `dismiss$.pipe(filter((s) => s === symbol))` returns an observable that doesn't complete, because `dismiss$` doesn't complete.
- On the other hand, the observable `timer(DISMISS_TIMEOUT)` emits once and completes immediately.
- Which means that the observable returned by `race` only completes if the `timer` wins the race. Because if the first observable wins, then that one doesn't complete.
- It is important that the observable returned by `race` always completes, because otherwise the inner-observable of `exhaustMap` won't complete. Keep in mind that `exhaustMap` is like the "opposite" of `switchMap`. In the sense that what `exhaustMap` does is to ignore the emissions of its source until its inner-observable hasn't completed.

There were other ways to accomplish the same. For instance:
- `dismiss$.pipe(takeUntil((s) => s === symbol))
- or
```ts
race([
  dismiss$.pipe(filter((s) => s === symbol)),
  timer(DISMISS_TIMEOUT),
]).pipe(take(1))
```
In fact, that that I wrote that, it's quite obvious to me that using `takeUntil` is a lot more declarative for this... So I will change that :slightly_smiling_face: 

Please @slcfv3 ask more questions if I didn't explain this properly or if this explanations leaves your with more doubts than answers. I do realize that this is a lot to digest.